core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1099 @@
+  });
+
+  it('should update the exploration version if it ' +
+    'is not in question player mode when initialized', fakeAsync(() => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    exploration.version = 2;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(false);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.callFake((expID, version) => {
+        expect(questionPlayerEngineService.getExplorationVersion()).toBe(1);
+        return Promise.resolve(exploration);
+      });
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    tick();
+
+    expect(questionPlayerEngineService.getExplorationVersion()).toBe(2);
+  }));
+
+  it('should update the current question ID when a new card is added', () => {
+    let successCallback = jasmine.createSpy('success');
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    let answer = 'answer';
+    let interactionRulesService: InteractionRulesService;
+    let answerClassificationResult = new AnswerClassificationResult(
+      outcomeObjectFactory
+        .createNew('default', '', '', []), 1, 0, 'default_outcome'
+    );
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+    let answerClassificationServiceSpy =
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => {
+        return html;
+      });
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    let currentQuestion1 = questionPlayerEngineService.getCurrentQuestion();
+    expect(currentQuestion1.getId()).toBe('questionId1');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+    let currentQuestion2 = questionPlayerEngineService.getCurrentQuestion();
+
+    expect(answerClassificationServiceSpy).toHaveBeenCalled();
```

Comment:
Why do you need this line (what is it checking and how does it relate to the test description)?

Note that, in general, checking whether a method has been called is a bit of an antipattern. That's because it tends to be directly testing implementation. The exception may be if you need to check that an endpoint method (e.g. to a third-party API which we treat as a black box) is called, but that's fairly rare.

Reply:
Previously I've added both test cases for testing implementation and interface.
Removing all redundant (implementation) test cases.

Reply:
Although I feel like we should test implementation to some extinct specifically for services, as they are not involved with UI directly.

Reply:
To be clear: here, "interface" does not refer to user interface. It refers to the part of the directive/service/component that can be accessed externally through public functions. So, services have an interface too (i.e. the public functions that they expose to other components/services).

Does that help clarify?

Reply:
Thanks for the info, removed!